File: scribbleshare-frontend/src/scripts/protocol/server/messages/ServerMessageCanvasMove.js
import ServerMessage from "../ServerMessage.js";
import ServerMessageType from "../ServerMessageType.js";
import CanvasMove from "../../../canvas/CanvasMove.js";

export default class ServerMessageCanvasMove extends ServerMessage {
    constructor(reader) {
        super(ServerMessageType.CANVAS_MOVE)
        this.canvasMovesMap = new Map();
        let length = reader.readUint8();
        for (let i = 0; i < length; i++) {
            let id = reader.readInt16();
            let canvasMoves = [];
            let lengthJ = reader.readUint8();
            for (let j = 0; j < lengthJ; j++) {
                canvasMoves.push(new CanvasMove(reader));
            }
            this.canvasMovesMap.set(id, canvasMoves);
        }
    }
}

File: scribbleshare-room/src/main/java/net/stzups/scribbleshare/room/server/websocket/protocol/server/messages/ServerMessageCanvasMove.java
package net.stzups.scribbleshare.room.server.websocket.protocol.server.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.scribbleshare.data.objects.canvas.CanvasMove;
import net.stzups.scribbleshare.room.server.websocket.protocol.server.ServerMessage;
import net.stzups.scribbleshare.room.server.websocket.protocol.server.ServerMessageType;

import java.util.Map;

public class ServerMessageCanvasMove extends ServerMessage {
    private final Map<Short, CanvasMove[]> canvasMovesMap;

    public ServerMessageCanvasMove(Map<Short, CanvasMove[]> canvasMovesMap) {
        super(ServerMessageType.CANVAS_MOVE);
        this.canvasMovesMap = canvasMovesMap;
    }

    @Override
    public void serialize(ByteBuf byteBuf) {
        super.serialize(byteBuf);
        byteBuf.writeByte((byte) canvasMovesMap.size());
        for (Map.Entry<Short, CanvasMove[]> entry : canvasMovesMap.entrySet()) {
            byteBuf.writeShort(entry.getKey());
            byteBuf.writeByte((byte) entry.getValue().length);
            for (CanvasMove canvasMove : entry.getValue()) {
                canvasMove.serialize(byteBuf);
            }
        }
    }
}
